Source/JavaScriptCore/builtins/JSIteratorConstructor.js

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2024 Sosuke Suzuki <aosukeke@gmail.com>.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// https://tc39.es/proposal-iterator-helpers/#sec-getiteratordirect
+@linkTimeConstant
+function getIteratorDirect(obj)
+{
+    "use strict";
+
+    // 1. Let nextMethod be ? Get(obj, "next").
+    var nextMethod = obj.next;
+    // 2. Let iteratorRecord be Record { [[Iterator]]: obj, [[NextMethod]]: nextMethod, [[Done]]: false }.
+    // 3. Return iteratorRecord.
+    return [obj, nextMethod];
+}
+
+// https://tc39.es/proposal-iterator-helpers/#sec-getiteratorflattenable
+@linkTimeConstant
+function getIteratorFlattenable(obj)
+{
+    "use strict";
+
+    // 1. If obj is not an Object, then
+    //   a. If stringHandling is reject-strings or obj is not a String, throw a TypeError exception.
+    if (!@isObject(obj) && typeof obj !== "string")
+        @throwTypeError("Iterafor.from requires an object or string");
+    // 2. Let method be ? GetMethod(obj, @@iterator).
+    var method = obj.@@iterator;
+    // 3. If method is undefined, then
+    //   a. Let iterator be obj.
+    var iterator;
+    if (@isUndefinedOrNull(method))
```

Comment:
Do,
```
if (@isUndefinedOrNull(obj.@@iterator))
```

Reply:
`obj.@@iterator` is used twice, so accepting your suggestion would break the following test case:

```js
{
    let symbolIteraterGetCount = 0;
    const iter = {
        get [Symbol.iterator]() {
          symbolIteraterGetCount++;
          return function() { return this };
        },
    };
    const wrapper = Iterator.from(iter);
    sameValue(symbolIteraterGetCount, 1);
}
```

So, this comment alone has not been addressed.